src/referenced-fields.ts

```diff
@@ -294,3 +313,271 @@ export const getModifiedFields: EngineInstance['getModifiedFields'] = (
 		return checkQuery(abstractSqlQuery);
 	}
 };
+
+type AliasedTable = {
+	tableName: string;
+	alias: string;
+};
+
+type BindCandidate = AliasedTable & {
+	scope: AbstractSqlQuery;
+};
+
+const findTablesInConstraint = (
+	constraint: BooleanTypeNodes,
+	candidates: BindCandidate[],
+) => {
+	// Recurse until we're sure that there are no more subqueries
+	switch (constraint[0]) {
+		case 'Boolean':
+			break;
+		case 'Equals':
+		case 'NotEquals':
+		case 'GreaterThan':
+		case 'GreaterThanOrEqual':
+		case 'LessThan':
+		case 'LessThanOrEqual':
+		case 'IsDistinctFrom':
+		case 'IsNotDistinctFrom':
+			const equalsNode = constraint as EqualsNode;
+			const left = equalsNode[1];
+			const right = equalsNode[2];
+			if (!isAbstractSqlQuery(left) || !isAbstractSqlQuery(right)) {
+				throw new Error(
+					'cannot introspect into the string form of AbstractSqlQuery',
+				);
+			}
+			if (SELECT_QUERY_TYPES.includes(left[0])) {
+				findBindCandidates(left, candidates);
```

Comment:
I think it would make sense for `findBindCandidates` to do:
```ts
if (!SELECT_QUERY_TYPES.includes(statement[0])) {
	return;
}
```
at the start rather than having to guard every call into it with this duplicated check

Reply:
What I mean is that instead of having
```ts
if (SELECT_QUERY_TYPES.includes(biOpLeft[0])) {
	findBindCandidates(biOpLeft, candidates);
}
if (SELECT_QUERY_TYPES.includes(biOpRight[0])) {
	findBindCandidates(biOpRight, candidates);
}
....
const findBindCandidates = (statement, candidates) => {
	...
}
```
that you have
```ts
findBindCandidates(biOpLeft, candidates);
findBindCandidates(biOpRight, candidates);
....
const findBindCandidates = (statement, candidates) => {
	if (!SELECT_QUERY_TYPES.includes(statement[0])) {
		return;
	}
	...
}
```
in order to reduce the duplication and remove the onus on the callsite by moving the parameter checking into the function itself

Reply:
Ah that makes sense :+1: